cluster-autoscaler/cloudprovider/clusterapi/clusterapi_machinedeployment.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterapi
+
+import (
+	"fmt"
+	"path"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/utils/pointer"
+)
+
+type machineDeploymentScalableResource struct {
+	controller        *machineController
+	machineDeployment *MachineDeployment
+	maxSize           int
+	minSize           int
+}
+
+var _ scalableResource = (*machineDeploymentScalableResource)(nil)
+
+func (r machineDeploymentScalableResource) ID() string {
+	return path.Join(r.Namespace(), r.Name())
+}
+
+func (r machineDeploymentScalableResource) MaxSize() int {
+	return r.maxSize
+}
+
+func (r machineDeploymentScalableResource) MinSize() int {
+	return r.minSize
+}
+
+func (r machineDeploymentScalableResource) Name() string {
+	return r.machineDeployment.Name
+}
+
+func (r machineDeploymentScalableResource) Namespace() string {
+	return r.machineDeployment.Namespace
+}
+
+func (r machineDeploymentScalableResource) Nodes() ([]string, error) {
+	var result []string
+
+	if err := r.controller.filterAllMachineSets(func(machineSet *MachineSet) error {
+		if machineSetIsOwnedByMachineDeployment(machineSet, r.machineDeployment) {
+			providerIDs, err := r.controller.machineSetProviderIDs(machineSet)
+			if err != nil {
+				return err
+			}
+			result = append(result, providerIDs...)
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func (r machineDeploymentScalableResource) Replicas() int32 {
+	return pointer.Int32PtrDerefOr(r.machineDeployment.Spec.Replicas, 0)
+}
+
+func (r machineDeploymentScalableResource) SetSize(nreplicas int32) error {
+	u, err := r.controller.dynamicclient.Resource(*r.controller.machineDeploymentResource).Namespace(r.machineDeployment.Namespace).Get(r.machineDeployment.Name, metav1.GetOptions{})
+
+	if err != nil {
+		return err
+	}
+
+	if u == nil {
+		return fmt.Errorf("unknown machineDeployment %s", r.machineDeployment.Name)
+	}
+
+	u = u.DeepCopy()
+	if err := unstructured.SetNestedField(u.Object, int64(nreplicas), "spec", "replicas"); err != nil {
```

Comment:
Why int64 here? I would expect int32 here.

Reply:
Under the hood `SetNestedField` eventually calls:

```go
// DeepCopyJSONValue deep copies the passed value, assuming it is a valid JSON representation i.e. only contains
// types produced by json.Unmarshal() and also int64.
// bool, int64, float64, string, []interface{}, map[string]interface{}, json.Number and nil
func DeepCopyJSONValue(x interface{}) interface{} {
}
```
Only `int64` is supported. Using `int32` panics:

```console
$ go test
panic: cannot deep copy int32 [recovered]
	panic: cannot deep copy int32

goroutine 1699 [running]:
testing.tRunner.func1(0xc000454200)
	/home/aim/go1.12.5/src/testing/testing.go:830 +0x392
panic(0x13b7e20, 0xc0004f1590)
	/home/aim/go1.12.5/src/runtime/panic.go:522 +0x1b5
k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/apimachinery/pkg/runtime.DeepCopyJSONValue(0x136a000, 0xc0006223b4, 0xc00030bbd0, 0x40b549)
	/home/aim/go-projects/autoscaler-merge/src/k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/apimachinery/pkg/runtime/converter.go:466 +0x572
k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.SetNestedField(0xc000406bd0, 0x136a000, 0xc0006223b4, 0xc00030bcc0, 0x2, 0x2, 0x0, 0x0)
	/home/aim/go-projects/autoscaler-merge/src/k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:209 +0x39
k8s.io/autoscaler/cluster-autoscaler/cloudprovider/clusterapi.machineDeploymentScalableResource.SetSize(0xc0002d7290, 0xc000359880, 0xa, 0x1, 0xc000000004, 0xc000359880, 0xc0003e23a8)
	/home/aim/go-projects/autoscaler-merge/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/clusterapi/clusterapi_machinedeployment.go:93 +0x1db
```

Reply:
Ah, gotcha...  makes sense even if it is a bit confusing given the structured types involved :joy: